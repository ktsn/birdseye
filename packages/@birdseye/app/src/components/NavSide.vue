<template>
  <ul class="nav-side">
    <li
      v-for="m in nav"
      :key="m.name"
      class="item"
    >
      <router-link
        :class="{ current: isCurrent(m.name) }"
        :to="previewRoute(m.name)"
        class="anchor"
      >
        {{ m.name }}
      </router-link>

      <TransitionDisclosure name="child">
        <div
          v-if="!isEmptyPatterns(m)"
          v-show="meta === m.name"
          class="child-wrapper"
        >
          <ul class="child-list">
            <li
              v-for="p in m.patterns"
              :key="p.name"
              class="child-item"
            >
              <router-link
                :class="{ current: isCurrent(m.name, p.name) }"
                :to="previewRoute(m.name, p.name)"
                class="child-anchor"
              >
                {{ p.name }}
              </router-link>
            </li>
          </ul>
        </div>
      </TransitionDisclosure>
    </li>
  </ul>
</template>

<script lang="ts">
import Vue from 'vue'
import { Location } from 'vue-router'
import { ComponentMeta } from '@birdseye/core'
import { TransitionDisclosure } from 'vue-transition-components'

export default Vue.extend({
  name: 'NavSide',

  components: {
    TransitionDisclosure
  },

  props: {
    meta: {
      type: String,
      default: null
    },

    pattern: {
      type: String,
      default: null
    },

    nav: {
      type: Array as () => ComponentMeta[],
      required: true
    }
  },

  methods: {
    previewRoute(meta: string, pattern?: string): Location {
      const to: Location = {
        name: 'preview',
        params: { meta }
      }

      if (pattern) {
        to.params!.pattern = pattern
      }

      return to
    },

    isCurrent(meta: string, pattern: string | null = null): boolean {
      return this.meta === meta && this.pattern === pattern
    },

    isEmptyPatterns(meta: ComponentMeta): boolean {
      return meta.patterns.length === 0
    }
  }
})
</script>

<style scoped>
.item {
  margin-bottom: 12px;
}

.item:last-child {
  margin-bottom: 0;
}

.anchor {
  display: block;
  font-weight: bold;
}

/**
 * Children
 */
.child-list {
  padding-top: 6px;
}

.child-item {
  margin-bottom: 6px;
}

.child-item:last-child {
  margin-bottom: 0;
}

.child-anchor {
  display: block;
  padding-left: 20px;
}

/**
 * Hover / Current
 */
.anchor:hover,
.anchor.current,
.child-anchor:hover,
.child-anchor.current {
  color: var(--color-main);
}

/**
 * Transition
 */
.child-enter-active,
.child-leave-active {
  transition: height 400ms var(--ease-out-cubic);
}
</style>
