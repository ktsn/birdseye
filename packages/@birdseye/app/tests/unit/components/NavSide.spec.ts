import Router from 'vue-router'
import { mount, createLocalVue } from '@vue/test-utils'
import NavSide from '@/components/NavSide.vue'

describe('NavSide', () => {
  const nav = [
    {
      name: 'Foo',
      patterns: [
        {
          name: 'Pattern 1'
        },
        {
          name: 'Pattern 2'
        }
      ]
    },
    {
      name: 'Bar',
      patterns: [
        {
          name: 'Pattern 1'
        }
      ]
    }
  ]

  let localVue: any

  beforeEach(() => {
    const router = new Router({
      routes: [
        {
          name: 'preview',
          path: '/:meta/:pattern?',
          component: {},
          props: true
        }
      ]
    })
    localVue = createLocalVue().extend({ router })
    localVue.use(Router)
  })

  it('renders declarations', () => {
    const wrapper = mount(NavSide, {
      localVue,
      propsData: {
        nav
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('highlight current component', () => {
    const wrapper = mount(NavSide, {
      localVue,
      propsData: {
        nav,
        meta: 'Foo'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('highlight current pattern', () => {
    const wrapper = mount(NavSide, {
      localVue,
      propsData: {
        nav,
        meta: 'Foo',
        pattern: 'Pattern 1'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('avoid including children dom when patterns are empty', () => {
    const wrapper = mount(NavSide, {
      localVue,
      propsData: {
        nav: [
          {
            name: 'Empty',
            patterns: []
          },
          {
            name: 'Empty Selected',
            patterns: []
          }
        ],
        meta: 'Empty Selected'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
