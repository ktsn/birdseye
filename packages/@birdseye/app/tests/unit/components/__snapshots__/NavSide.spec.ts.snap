// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavSide avoid including children dom when patterns are empty 1`] = `
<ul class="nav-side">
  <li class="item"><button class="anchor">
      Empty
    </button>
    <!---->
  </li>
  <li class="item"><button class="anchor current">
      Empty Selected
    </button>
    <!---->
  </li>
</ul>
`;

exports[`NavSide highlight current component 1`] = `
<ul class="nav-side">
  <li class="item"><button class="anchor current">
      Foo
    </button>
    <div class="child-wrapper">
      <ul class="child-list">
        <li class="child-item"><a href="#/Foo/Pattern%201" class="child-anchor">
            Pattern 1
          </a></li>
        <li class="child-item"><a href="#/Foo/Pattern%202" class="child-anchor">
            Pattern 2
          </a></li>
      </ul>
    </div>
  </li>
  <li class="item"><button class="anchor">
      Bar
    </button>
    <div class="child-wrapper" style="display: none;">
      <ul class="child-list">
        <li class="child-item"><a href="#/Bar/Pattern%201" class="child-anchor">
            Pattern 1
          </a></li>
      </ul>
    </div>
  </li>
</ul>
`;

exports[`NavSide highlight current pattern 1`] = `
<ul class="nav-side">
  <li class="item"><button class="anchor current">
      Foo
    </button>
    <div class="child-wrapper">
      <ul class="child-list">
        <li class="child-item"><a href="#/Foo/Pattern%201" class="child-anchor current">
            Pattern 1
          </a></li>
        <li class="child-item"><a href="#/Foo/Pattern%202" class="child-anchor">
            Pattern 2
          </a></li>
      </ul>
    </div>
  </li>
  <li class="item"><button class="anchor">
      Bar
    </button>
    <div class="child-wrapper" style="display: none;">
      <ul class="child-list">
        <li class="child-item"><a href="#/Bar/Pattern%201" class="child-anchor">
            Pattern 1
          </a></li>
      </ul>
    </div>
  </li>
</ul>
`;

exports[`NavSide renders declarations 1`] = `
<ul class="nav-side">
  <li class="item"><button class="anchor">
      Foo
    </button>
    <div class="child-wrapper" style="display: none;">
      <ul class="child-list">
        <li class="child-item"><a href="#/Foo/Pattern%201" class="child-anchor">
            Pattern 1
          </a></li>
        <li class="child-item"><a href="#/Foo/Pattern%202" class="child-anchor">
            Pattern 2
          </a></li>
      </ul>
    </div>
  </li>
  <li class="item"><button class="anchor">
      Bar
    </button>
    <div class="child-wrapper" style="display: none;">
      <ul class="child-list">
        <li class="child-item"><a href="#/Bar/Pattern%201" class="child-anchor">
            Pattern 1
          </a></li>
      </ul>
    </div>
  </li>
</ul>
`;
